finetune/util/metrics.py

```diff
@@ -120,61 +121,122 @@ def sequence_labeling_token_counts(true, predicted):
     return d
 
 
-def seq_recall(true, predicted, count_fn):
+def calc_recall(TP, FN):
+    try:
+        return TP / float(FN + TP)
+    except ZeroDivisionError:
+        return 0.0
+
+
+def calc_precision(TP, FP):
+    try:
+        return TP / float(FP + TP)
+    except ZeroDivisionError:
+        return 0.0
+
+
+def calc_f1(recall, precision):
+    try:
+        return 2 * (recall * precision) / (recall + precision)
+    except ZeroDivisionError:
+        return 0.0
+
+
+def seq_recall(true, predicted, span_type="token"):
+    count_fn = get_seq_count_fn(span_type)
     class_counts = count_fn(true, predicted)
     results = {}
     for cls_, counts in class_counts.items():
         FN = len(counts['false_negatives'])
         TP = len(counts['correct'])
-        try:
-            results[cls_] = TP / float(FN + TP)
-        except ZeroDivisionError: 
-            results[cls_] = 0.
+        results[cls_] = calc_recall(TP, FN)
     return results
 
 
-def seq_precision(true, predicted, count_fn):
+def seq_precision(true, predicted, span_type="token"):
+    count_fn = get_seq_count_fn(span_type)
     class_counts = count_fn(true, predicted)
     results = {}
     for cls_, counts in class_counts.items():
         FP = len(counts['false_positives'])
         TP = len(counts['correct'])
-        try:
-            results[cls_] = TP / float(FP + TP)
-        except ZeroDivisionError:
-            results[cls_] = 0.
+        results[cls_] = calc_precision(TP, FP)
     return results
 
-def micro_f1(true, predicted, count_fn):
+
+def micro_f1(true, predicted, span_type="token"):
+    count_fn = get_seq_count_fn(span_type)
     class_counts = count_fn(true, predicted)
     TP, FP, FN = 0, 0, 0
     for cls_, counts in class_counts.items():
         FN += len(counts['false_negatives'])
         TP += len(counts['correct'])
         FP += len(counts['false_positives'])
-    recall = TP/float(FN + TP)
-    precision = TP / float(FP + TP)
-    f1 = 2 * (recall * precision) / (recall + precision)
-    return f1
+    recall = calc_recall(TP, FN)
+    precision = calc_precision(TP, FP)
+    return calc_f1(recall, precision)
+
+
+def per_class_f1(true, predicted, span_type="token"):
+    """
+    F1-scores per class
+    """
+    count_fn = get_seq_count_fn(span_type)
+    class_counts = count_fn(true, predicted)
+    results = {}
+    for cls_, counts in class_counts.items():
+        results[cls_] = {}
+        FP = len(counts["false_positives"])
+        FN = len(counts["false_negatives"])
```

Comment:
One final request -- could we call this 'true_positives'?  I think "correct" was a terrible naming convention I wrought upon metrics.py a long ago

Reply:
updated